frontend/src/components/c-file-type-checkbox.vue

```diff
@@ -0,0 +1,124 @@
+<template lang="pug">
+.checkboxes.mui-form--inline(v-if="fileTypes.length > 0")
+  label.all-checkbox
+    input.mui-checkbox--fileType#all(type="checkbox", v-model="isAllChecked", value="all")
+    span(v-if="allCheckboxLabel", v-bind:title="getTitle(allCheckboxLabel)")
+      span {{ getLabel(allCheckboxLabel) }}
+    span(v-else) All&nbsp;
+  label(
+    v-for="fileType, index in fileTypes",
+    v-bind:key="fileType",
+    v-bind:style="{\
+      'background-color': fileTypeColors[fileType],\
+      'color': getFontColor(fileTypeColors[fileType])\
+      }"
+  )
+    input.mui-checkbox--fileType(type="checkbox", v-bind:value="fileType",
+      v-model="localSelectedFileTypes", v-bind:id="fileType")
+    span(v-if="fileTypeCheckboxLabels", v-bind:title="getTitle(fileTypeCheckboxLabels[index])")
+      span {{ getLabel(fileTypeCheckboxLabels[index]) }}
+    span(v-else) {{ this.fileTypes[index] }}&nbsp;
+</template>
+
+<script lang="ts">
+import { defineComponent, PropType } from 'vue';
+
+export default defineComponent({
+  name: 'c-file-type-checkbox',
+  props: {
+    fileTypes: {
+      type: Array as PropType<Array<string>>,
+      required: true,
+    },
+    fileTypeColors: {
+      type: Object as PropType<Record<string, string>>,
+      required: true,
+    },
+    selectedFileTypes: {
+      type: Array as PropType<Array<string>>,
+      required: true,
+    },
+    allCheckboxLabel: {
+      type: Object as PropType<{
+        fileTitle: string,
+        fileType: string,
+        lineCount: number,
+        blankLineCount: number,
+      }>,
+      default: undefined,
+    },
+    fileTypeCheckboxLabels: {
+      type: Array as PropType<Array<{
+        fileTitle: string,
+        fileType: string,
+        lineCount: number,
+        blankLineCount: number,
+      }>>,
+      default: undefined,
+    },
+  },
+  emits: ['update-selected-file-types-hash', 'update:selectedFileTypes', 'select-all-checked'],
+  computed: {
+    isAllChecked: {
+      get(): boolean {
+        return this.selectedFileTypes.length === this.fileTypes.length;
+      },
+      set(value: boolean): void {
+        if (value) {
+          this.localSelectedFileTypes = this.fileTypes.slice();
+        } else {
+          this.localSelectedFileTypes = [];
+        }
+        this.$emit('select-all-checked', value);
+      },
+    },
+    localSelectedFileTypes: {
+      get(): Array<string> {
+        return this.selectedFileTypes;
+      },
+      set(value: Array<string>): void {
+        this.$emit('update:selectedFileTypes', value);
+        this.updateSelectedFileTypesHash();
+      },
+    },
+  },
+  methods: {
+    getFontColor(color: string): string {
+      return window.getFontColor(color);
+    },
+    updateSelectedFileTypesHash(): void {
+      this.$emit('update-selected-file-types-hash');
+    },
```

Comment:
Is there a need to differentiate the `update-selected-file-types-hash` and `update:selectedFileTypes` emits? I personally feel that it might be okay to remove the `update-selected-file-types-hash` emit since this component doesn't need to know about the concept of hashes at all - it can just emit `update:selectedFileTypes` whenever the selectedFileTypes is updated/changed, and let the parent component decide what to do in response to that (which might include updating the hash).

For example, in c-summary, we listen to the emit and run the `getFiltered` method, but we run it _in repsonse to the checkedFileTypes changing_, rather than in response to needing to update the selected file types hash (updating the hash is one of the things done in `getFiltered`, among others).

Reply:
I agree that it would be better to hide this interaction from the component! I have updated the code.